File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    const check = (node, m, n) => {
        if (!node) return true;
        if (node.val >= m || node.val <= n) return false;
        const left = check(node.left, node.val, n);
        const right = check(node.right, m, node.val);
        return left && right;
    }
    return check(root, Infinity, -Infinity);
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def check(node, m, n):
            if not node:
                return True
            if (node.val >= m) or (node.val <= n):
                return False
            left = check(node.left, node.val, n)
            right = check(node.right, m, node.val)
            return left and right
        return check(root, float('inf'), float('-inf'))